src/main/java/io/supertokens/jwt/JWTSigningKey.java

```diff
@@ -0,0 +1,247 @@
+
+            while (true) {
+                List<JWTSigningKeyInfo> keysFromStorage = noSQLStorage.getJWTSigningKeys_Transaction();
+
+                // If there are no keys in storage, generate a new one
+                if (keysFromStorage.isEmpty()) {
+                    try {
+                        keyInfo = generateKeyForAlgorithm(algorithm);
+                        boolean success = noSQLStorage.setJWTSigningKeyInfo_Transaction(keyInfo);
+
+                        if (!success) {
+                            continue;
+                        }
+
+                        return keyInfo;
+                    } catch (NoSuchAlgorithmException e) {
+                        throw new StorageTransactionLogicException(e);
+                    }
+                } else {
+                    // Loop through the keys and find the first one for the algorithm
+                    for (int i = 0; i < keysFromStorage.size(); i++) {
+                        JWTSigningKeyInfo currentKey = keysFromStorage.get(i);
+                        if (currentKey.algorithm.equalsIgnoreCase(algorithm.name())) {
+                            keyInfo = currentKey;
+                            break;
+                        }
+                    }
+
+                    // If no key was found create a new one
+                    if (keyInfo == null) {
+                        try {
+                            keyInfo = generateKeyForAlgorithm(algorithm);
+                            boolean success = noSQLStorage.setJWTSigningKeyInfo_Transaction(keyInfo);
+
+                            if (!success) {
+                                continue;
+                            }
+
+                            return keyInfo;
+                        } catch (NoSuchAlgorithmException e) {
+                            throw new StorageTransactionLogicException(e);
+                        }
+                    } else {
+                        return keyInfo;
+                    }
+                }
+            }
+        }
+
+        throw new QuitProgramException("Unsupported storage type detected");
+    }
+
+    private JWTSigningKeyInfo generateKeyForAlgorithm(SupportedAlgorithms algorithm) throws NoSuchAlgorithmException,
+            UnsupportedJWTSigningAlgorithmException {
+        if (algorithm.getAlgorithmType().equalsIgnoreCase("rsa")) {
+            long currentTimeInMillis = System.currentTimeMillis();
+            Utils.PubPriKey newKey = Utils.generateNewPubPriKey();
+            return new JWTAsymmetricSigningKeyInfo(Utils.getUUID(), currentTimeInMillis, algorithm.name(), newKey.publicKey,
+                    newKey.privateKey);
+        }
+
+        // Ideally this should never execute because for unsupported algorithms an exception will be thrown before this method
+        // This is just so that we dont need to handle null when we call this method
+        throw new UnsupportedJWTSigningAlgorithmException();
```

Comment:
This is not good logic. If someone passes null into this function, throw some in built runtime exception and not this specific excption that we have. I believe `UnsupportedJWTSigningAlgorithmException` is thrown when creating the `SupportedAlgorithms` object?

Reply:
Throwing IllegalArgument as discussed